org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/execution/ExecutionPage.java

```diff
@@ -313,24 +316,37 @@ private ActionShowingContextMenuListener createContextMenuListener(TreeViewer tr
         RunTestAction runTestAction = new RunTestAction(this);
         ShowFailureAction showFailureAction = new ShowFailureAction(this);
         OpenTestSourceFileAction openTestSourceFileAction = new OpenTestSourceFileAction(this);
-        ExpandTreeNodesAction expandNodesAction = new ExpandTreeNodesAction(treeViewer);
-        CollapseTreeNodesAction collapseNodesAction = new CollapseTreeNodesAction(treeViewer);
 
-        List<SelectionSpecificAction> contextMenuActions = ImmutableList.<SelectionSpecificAction>of(runTestAction, showFailureAction, openTestSourceFileAction, expandNodesAction, collapseNodesAction);
-        List<SelectionSpecificAction> contextMenuActionsPrecededBySeparator = ImmutableList.<SelectionSpecificAction>of(openTestSourceFileAction, expandNodesAction);
+        List<SelectionSpecificAction> contextMenuActions = ImmutableList.<SelectionSpecificAction>of(runTestAction, showFailureAction, openTestSourceFileAction);
+
+        List<SelectionSpecificAction> contextMenuActionsPrecededBySeparator = ImmutableList.<SelectionSpecificAction>of(openTestSourceFileAction);
         ImmutableList<SelectionSpecificAction> contextMenuActionsSucceededBySeparator = ImmutableList.of();
 
         return new ActionShowingContextMenuListener(this, contextMenuActions, contextMenuActionsPrecededBySeparator, contextMenuActionsSucceededBySeparator);
     }
 
     private void registerListeners() {
-        // navigate to source file on double click or when pressing enter
+        // navigate to source file or expand or collapse group node on double click or when pressing enter
         getPageControl().getViewer().addDoubleClickListener(new IDoubleClickListener() {
 
             @Override
             public void doubleClick(DoubleClickEvent event) {
+                NodeSelection nodeSelection = NodeSelection.from(event.getSelection());
                 OpenTestSourceFileAction openTestSourceFileAction = new OpenTestSourceFileAction(ExecutionPage.this);
-                openTestSourceFileAction.run();
+
+                if (openTestSourceFileAction.isVisibleFor(nodeSelection) && openTestSourceFileAction.isEnabledFor(nodeSelection)) {
+                    openTestSourceFileAction.run();
+                } else if (nodeSelection.isSingleSelection()) {
+                    Object selected = nodeSelection.toList().get(0);
+                    IContentProvider provider = getPageControl().getViewer().getContentProvider();
```

Comment:
For readability, `getPageControl().getViewer()` should be extracted to local variable.

Reply:
done